### PR TITLE
Null check return advices

### PR DESCRIPTION
**Pull request by SylvainJuge**
## What does this PR do?

For exit method advices:
- thrown exception can be null or not (if an exception has been thrown or not)
- return value can be null when an exception is thrown

Thus as a good practice all exit method advices should:
- have BOTH of those parameters annotated with `@Nullable`
- ensure proper handling of those null values

This PR:
- adds missing annotations on advice parameters
- adds a few extra missing null checks

## Checklist

- [x] This is a bugfix
  - [x] I have updated [CHANGELOG.asciidoc](https://github.com/elastic/apm-agent-java/blob/master/CHANGELOG.asciidoc)
  - [ ] I have added tests that would fail without this fix

### CODE REVIEW

File: apm-agent-plugins/apm-dubbo-plugin/src/main/java/co/elastic/apm/agent/dubbo/advice/AlibabaMonitorFilterAdvice.java [hunk 1]
```diff
@@ -87,17 +87,21 @@ public static void onEnterFilterInvoke(@Advice.Argument(1) Invocation invocation
...
+                                          @Advice.Return @Nullable Result result,
+                                          @Advice.Local("span") @Nullable Span span,
+                                          @Advice.Thrown @Nullable Throwable t,
+                                          @Advice.Local("transaction") @Nullable Transaction transaction) {
         AbstractSpan<?> actualSpan = span != null ? span : transaction;
         if (actualSpan == null) {
             return;
         }
-        actualSpan.captureException(t)
-            .captureException(result.getException())
```
Comment (SylvainJuge) [inline apm-agent-plugins/apm-dubbo-plugin/src/main/java/co/elastic/apm/agent/dubbo/advice/AlibabaMonitorFilterAdvice.java:99]: fixed potential NPE when result was null

File: apm-agent-plugins/apm-grpc/apm-grpc-plugin/src/main/java/co/elastic/apm/agent/grpc/ServerCallHandlerInstrumentation.java [hunk 2]
```diff
@@ -79,7 +79,11 @@ public static void onExit(@Advice.Thrown @Nullable Throwable thrown,
             return;
         }
 
-        helper.registerTransaction(serverCall, listener, transaction);
```
Comment (SylvainJuge) [inline apm-agent-plugins/apm-grpc/apm-grpc-plugin/src/main/java/co/elastic/apm/agent/grpc/ServerCallHandlerInstrumentation.java:82]: fixed potential NPE when listener is null

File: apm-agent-plugins/apm-jdbc-plugin/src/main/java/co/elastic/apm/agent/jdbc/ConnectionInstrumentation.java [hunk 3]
```diff
@@ -50,9 +51,12 @@
 public class ConnectionInstrumentation extends JdbcInstrumentation {
 
     @Advice.OnMethodExit(suppress = Throwable.class, inline = false)
-    public static void storeSql(@Advice.Return PreparedStatement statement,
+    public static void storeSql(@Advice.Return @Nullable PreparedStatement statement,
                                 @Advice.Argument(0) String sql) {
-        getJdbcHelper().mapStatementToSql(statement, sql);
```
Comment (SylvainJuge) [inline apm-agent-plugins/apm-jdbc-plugin/src/main/java/co/elastic/apm/agent/jdbc/ConnectionInstrumentation.java:55]: fixed potential NPE when statement is null (as map does not allow for null values)

File: apm-agent-plugins/apm-jdk-httpclient-plugin/src/main/java/co/elastic/apm/agent/httpclient/HttpClientAsyncInstrumentation.java [hunk 4]
```diff
@@ -77,6 +77,8 @@ public static void onAfterExecute(@Advice.Return @Nullable CompletableFuture<Htt
             if (completableFuture == null) {
                 span.captureException(t)
                     .end();
+
+                return;
```
Comment (SylvainJuge) [inline apm-agent-plugins/apm-jdk-httpclient-plugin/src/main/java/co/elastic/apm/agent/httpclient/HttpClientAsyncInstrumentation.java:81]: fixed potential NPE when completableFuture is null

File: apm-agent-plugins/apm-jdk-httpclient-plugin/src/main/java/co/elastic/apm/agent/httpclient/HttpRequestHeadersInstrumentation.java [hunk 5]
```diff
@@ -60,12 +61,13 @@
...
 
-    @Nonnull
+
+    @Nullable
     @AssignTo.Return
     @Advice.OnMethodExit(suppress = Throwable.class, onThrowable = Throwable.class, inline = false)
-    public static HttpHeaders onAfterExecute(@Advice.Return @Nonnull final HttpHeaders httpHeaders) {
+    public static HttpHeaders onAfterExecute(@Advice.Return @Nullable final HttpHeaders httpHeaders) {
         Span span = tracer.getActiveSpan();
-        if (span == null) {
```
Comment (SylvainJuge) [inline apm-agent-plugins/apm-jdk-httpclient-plugin/src/main/java/co/elastic/apm/agent/httpclient/HttpRequestHeadersInstrumentation.java:68]: fixed potential NPE when httpHeaders is null

File: apm-agent-plugins/apm-process-plugin/src/main/java/co/elastic/apm/agent/process/ProcessStartInstrumentation.java [hunk 6]
```diff
@@ -74,7 +73,12 @@ public static void onExit(@Advice.This ProcessBuilder processBuilder,
                 parentSpan.captureException(t);
             }
 
-            ProcessHelper.startProcess(parentSpan, process, processBuilder.command());
```
Comment (SylvainJuge) [inline apm-agent-plugins/apm-process-plugin/src/main/java/co/elastic/apm/agent/process/ProcessStartInstrumentation.java:77]: fixed NPE

File: apm-agent-plugins/apm-servlet-plugin/src/main/java/co/elastic/apm/agent/servlet/AsyncInstrumentation.java [hunk 7]
```diff
@@ -108,7 +108,10 @@
             private static final AsyncContextAdviceHelper<AsyncContext> asyncHelper = new AsyncContextAdviceHelperImpl(GlobalTracer.requireTracerImpl());
 
             @Advice.OnMethodExit(suppress = Throwable.class, inline = false)
-            public static void onExitStartAsync(@Advice.Return AsyncContext asyncContext) {
```
Comment (SylvainJuge) [inline apm-agent-plugins/apm-servlet-plugin/src/main/java/co/elastic/apm/agent/servlet/AsyncInstrumentation.java:111]: fixed potential NPE, for example when trying to start an asynchronous servlet that does not support async (very very unlikely in production).

**Review by eyalkoren — APPROVED** [review]

**Pull request merged**